compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -324,18 +328,50 @@ impl<'f> PerFunctionContext<'f> {
         self.blocks.insert(block, references);
     }
 
-    /// Add a self-alias for input parameters, similarly to how a newly allocated reference has
-    /// one alias already - itself. If we don't, then the checks using `reference_parameters()`
-    /// might find the default (empty) aliases and think the an input reference can be removed.
+    /// Go through each parameter and register that all reference parameters of the same type are
+    /// possibly aliased to each other. If there are parameters with nested references (arrays of
+    /// references or references containing other references) we give up and assume all parameter
+    /// references are `AliasSet::unknown()`.
     fn add_aliases_for_reference_parameters(&self, block: BasicBlockId, references: &mut Block) {
         let dfg = &self.inserter.function.dfg;
         let params = dfg.block_parameters(block);
-        let params = params.iter().filter(|p| dfg.value_is_reference(**p));
+
+        let mut aliases: HashMap<Type, AliasSet> = HashMap::default();
 
         for param in params {
-            let expression =
-                references.expressions.entry(*param).or_insert(Expression::Other(*param));
-            references.aliases.entry(expression.clone()).or_insert_with(|| AliasSet::known(*param));
+            match dfg.type_of_value(*param) {
+                // If the type indirectly contains a reference we have to assume all references
+                // are unknown since we don't have any ValueIds to use.
+                Type::Reference(element) if element.contains_reference() => return,
+                Type::Reference(element) => {
+                    let empty_aliases = AliasSet::known_empty();
+                    let alias_set =
+                        aliases.entry(element.as_ref().clone()).or_insert(empty_aliases);
+                    alias_set.insert(*param);
+                }
+                typ if typ.contains_reference() => return,
+                _ => continue,
+            }
+        }
+
+        for aliases in aliases.into_values() {
+            if let Some(alias) = aliases.single_alias() {
+                let expression =
+                    references.expressions.entry(alias).or_insert(Expression::Other(alias));
+                references
+                    .aliases
+                    .entry(expression.clone())
+                    .or_insert_with(|| AliasSet::known(alias));
+            } else if let Some(first) = aliases.first() {
+                let expression = Expression::Other(first);
+                let previous = references.aliases.insert(expression.clone(), aliases.clone());
+                assert!(previous.is_none());
+
+                aliases.for_each(|alias| {
+                    let previous = references.expressions.insert(alias, expression.clone());
+                    assert!(previous.is_none());
+                });
```

Comment:
Do we need the first branch here? Seems like the second can cover both situations (also we can assert that `aliases.first()` is `Some` as we should never end up with an empty or unknown `AliasSet` in this iter.

Reply:
I think you're right. Since we expect parameters to not already be in each map the `entry` versus `insert` distinction isn't needed and the only difference afterward is some inexpensive cloning of `expression`. I'll update it.